If a more recent revision/autosave exists, store its state on editor setup

talldan opened the pull request:
closes #7416

Attempts to resolve an issue whereby the editor is unaware of the existence of
a more recent autosave, resulting in attempts to perform an autosave fail.

## Description
This PR results in autosave state being fetched when the editor loads (via a resolver). This request is preloaded (will require a core patch). Changes include:
- Migration of autosave state from '@wordpress/editor' to '@wordpress/core-data'.
- Addition of resolver for 'getAutosave' selector.
- Disabling of the preview button until autosave state is fetched.
- Deprecation of old selectors/actions

## How has this been tested?
- Manual testing - follow the steps in #7416
- Added unit tests

## Types of changes
<!-- What types of changes does your code introduce?  -->
Bug fix (non-breaking change which fixes an issue)
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->

aduth commented on lib/client-assets.php:
```diff
@@ -1216,6 +1216,9 @@ function gutenberg_editor_scripts_and_styles( $hook ) {
 	if ( $post_autosave ) {
 		$editor_settings['autosave'] = array(
 			'editLink' => add_query_arg( 'gutenberg', true, get_edit_post_link( $post_autosave->ID ) ),
+			'title'    => $post_autosave->post_title,
```
This does not take the shape nor apply the same filtering as is done in the REST revisions controller which is otherwise used by the application, and probably should.

https://github.com/WordPress/WordPress/blob/4c2fcf36a3193a7e63a82e5ff459f2587c69a859/wp-includes/rest-api/endpoints/class-wp-rest-revisions-controller.php#L386-L407

Maybe `get_autosave_newer_than_post_save` should be returning a REST entity? Seems like we have two shapes of an autosave between how we're bootstrapping vs. otherwise using in the editor, and I don't see why it ought to be that way.

talldan replied:
Would it be best to extract the building of the autosave object from the rest api code? Maybe relocate it to the revisions.php file?

aduth replied:
The [`gutenberg_preload_api_request` function](https://github.com/WordPress/gutenberg/blob/b06a33569a2b860110cb8d3f113ccf6c2b36e492/lib/client-assets.php#L566-L618) could probably server as prior art here.

talldan commented:
Hey @aduth - I've pushed some further changes as discussed. I've noticed these fixes don't quite resolve things, there's a further issue.

The logic for checking whether an autosave should be made for the REST api is comparing the attempted autosave against the post content:
https://github.com/WordPress/gutenberg/blob/master/lib/class-wp-rest-autosaves-controller.php#L313

But the frontend is comparing the attempted autosave against the previous autosave:
https://github.com/WordPress/gutenberg/blob/master/editor/store/selectors.js#L385

Because these two don't match up the `There is nothing to save...` error is still happening when attempting to preview a post if that post has a later autosave and the content hasn't changed.

I think the right way to go to resolve this is to bring the rest api and the front-end in line. I think they should both check the latest saved post content and the autosave content to determine if the post can be autosaved. What do you think?

youknowriad commented:
This seems an important bug to fix, but I'm moving to 3.8 for now as it doesn't seem ready.

talldan commented:
I've pushed a new commit that I think actually fixes the issue.

The logic I've add means that a post can only be autosaved if it's different to the latest save of the post and the current autosave (when one exists that's newer that the last save of the post).

I think that makes sense, would be great to get a re-review!

talldan commented:
@aduth Apologies this one has been open for so long - it should be good for a re-review if you get a moment.

aduth commented:
There's quite a bit which has since been changed in affected code here, for which I'm largely to blame (both for the changes and neglecting to review this sooner). Let me know if there's anything I can do to help with bringing this back up to date.

talldan commented:
@aduth - I've rebased it. I ended up dropping a couple of commits that didn't seem relevant any more. They seem to have been fixed since.

My understanding is that if we want to introduce this, the PHP code will need to patched into core. Is that right?

aduth commented:
Yes, and there may be some conflict to anticipate between this and the changes introduced to core as part of [Trac#45194](https://core.trac.wordpress.org/ticket/45194) (specifically [43833](https://core.trac.wordpress.org/changeset/43833)), which doesn't appear to have yet been ported back to the plugin (cc @imath @danielbachhuber).

danielbachhuber commented:
> the changes introduced to core as part of [Trac#45194](https://core.trac.wordpress.org/ticket/45194) (specifically [43833](https://core.trac.wordpress.org/changeset/43833)), which doesn't appear to have yet been ported back to the plugin

It's still a WIP on #4155

aduth commented:
> > the changes introduced to core as part of [Trac#45194](https://core.trac.wordpress.org/ticket/45194) (specifically [43833](https://core.trac.wordpress.org/changeset/43833)), which doesn't appear to have yet been ported back to the plugin
> 
> It's still a WIP on #4155

Okay, it'd be easier all things considered if it were made to be its own pull request, but I'll plan to give #4155 a look.

aduth commented on lib/client-assets.php:
```diff
@@ -1652,8 +1668,11 @@ function gutenberg_editor_scripts_and_styles( $hook ) {
 	if ( $post_autosave ) {
-		$editor_settings['autosave'] = array(
-			'editLink' => add_query_arg( 'gutenberg', true, get_edit_post_link( $post_autosave->ID ) ),
+		$editor_settings['autosave'] = array_merge(
```
Looking more, I'm more inclined to think that anything which _can_ be fulfilled in the client by mere awareness of the edited post's IDs _should be_ (preloading fine to offset load delay), since it's not really a "setting" in the true sense of the word.

If it comes to it, I'm willing to relent to let this pass through in its current form though.

aduth replied:
The general approach would be to show the most we can given the data we know. If we need to know the availability of an autosave before a preview can be performed, we disable or otherwise prevent the user from being able to initiate that action. Or, in the very, very worst case (i.e. bad), we don't show the editor at all until all requisite data is known. This technically already occurs for having the post data to edit. We remedy the asynchronous nature of data availability by [preloading data](https://github.com/WordPress/gutenberg/blob/fba803e46ba930b09e3763d224e5abfdf0275ed2/lib/client-assets.php#L1426-L1434) such that a network request which would normally take place is preempted to defer to data made available by the server at load-time.

The indirectness of this solution is why I'd made a point to mention my willingness to relent 😄 But in general, the minimal necessary to edit a post (ideally its ID) is all we'd provide from the server, and the rest is issued as a dependency thereof within the client itself.

talldan replied:
Oh right! I understand now—of course it can be preloaded, sorry to make you mention that twice.

I've refactored this to use apiFetch and have preloaded autosave data. That does make this a much more minimal change which is nice. I wasn't sure if an effect was the right way to do this, but that's how it works at the moment.

talldan committed:
Refactor autosave state into core-data

- Migrate state from core/editor to core
- Adjust selectors and action creators to take a post id.
- Store autosave state by the post id.
- Deprecate old action creators and selectors and forward to new implementation
- Update tests
- Update changelog

Add a resolver for fetching autosave data

Preload any autosaves for the current post

Improve isEditedPostAutosaveable selector deprecation

Pass entire parent post object into autosave selector functions to allow the getAutosave resolver to form a REST API path that incorporates both the post id and type

Use a selector to get the correct post type base url for the autosave

Rename resetAutosave to receiveAutosave for conformity

Render the post preview button as disabled until autosave has been fetched.

Comments/test description tidy up

Prefer `action` over `action creator` in copy

Remove unecessary if statement

Remove `autosave` prop overriding the prop of the same name in withDispatch

Fix deprecated resetAutosave function

Fix jsdoc formatting

Remove new getAutosaveAttribute in favour of deprecating and removing old version

Change autosave selector interface from func( post ) to func( postType, postId )

Store raw autosave from REST API in core, move transformation of autosave back to the core/editor package

Remove and deprecate hasAutosave in favour of getAutosave

Update docs

Rename reducer autosave -> autosaves

Fix erroneous changelog entry caused by bad merge conflict resolution

Update deprecations

Remove unnecessary new line

Use createRegistrySelector to select autosave data from correct registry for deprecated functions

Handle usage of registry selector in tests for isEditedPostAutosaveable

aduth reviewed (changes requested):
It's going in a nice direction with the recent revisions.

A few things I'd like to see:

- An effect like the one we've added here with complementing "fetch" action is being gradually deprecated in favor of `@wordpress/data` resolvers and specifically the [Core Data entities](https://github.com/WordPress/gutenberg/tree/master/packages/core-data).
   - The editor's post initialization already [occurs this way](https://github.com/WordPress/gutenberg/blob/9dfb43c0360d049e75dedf4e879bba893634424b/packages/edit-post/src/editor.js#L51), and can serve as prior art in how it's a [blocker for rendering the editor](https://github.com/WordPress/gutenberg/blob/9dfb43c0360d049e75dedf4e879bba893634424b/packages/edit-post/src/editor.js#L21-L23).
   - Better yet, moving this up to being resolved by `edit-post/editor` allows the request to occur _in parallel with_, rather than in series by, the post "request"
- There's nothing with the implementation here which would prevent the user from quickly pressing preview if the preloading was not in place.
   - Considering the preloading to be treated as an add-on enhancement, not an assumed thing, can help identify some of these potential UX issues.

talldan commented on packages/editor/src/store/resolvers.js:
```diff
@@ -0,0 +1,19 @@
+ *
+ * @param {string} postId The Post to retrieve the autosave for.
+ */
+export function* getAutosave( postId ) {
```
It seems a bit unusual that the getAutosave resolver requires a postId parameter, but the selector doesn't. Should the autosave state be keyed by postId? That seems unnecessary, but I'm not sure how else to resolve the differences.

talldan commented on packages/editor/src/store/index.js:
```diff
@@ -7,9 +7,11 @@ import { registerStore } from '@wordpress/data';
  * Internal dependencies
  */
 import reducer from './reducer';
+import controls from './controls';
```
Controls is boilerplate that would be good to remove.

talldan commented on packages/editor/src/store/resolvers.js:
```diff
@@ -0,0 +1,19 @@
+export function* getAutosave( postId ) {
+	const autosaveResponse = yield apiFetch( { path: `/wp/v2/posts/${ postId }/autosaves?context=edit` } );
+	if ( autosaveResponse && autosaveResponse[ 0 ] ) {
+		yield resetAutosave( autosaveResponse[ 0 ] );
```
One issue to solve is that the autosave should only be stored when it's newer than the current post's modified date, so that it matches this bit of logic - https://github.com/WordPress/gutenberg/blob/master/lib/client-assets.php#L1276

talldan commented on packages/editor/src/components/post-preview-button/index.js:
```diff
@@ -197,15 +197,18 @@ export default compose( [
 		return {
-			postId: getCurrentPostId(),
+			postId,
+			autosave: getAutosave( postId ),
```
I'm not actually using this to disable use of the preview button yet. That's because it's possible for a post to not have an autosave and the preview button should still be usable. I think there'll need to be some separate state like 'hasCheckedForAutosave' that is stored, and when that's false the preview button is disabled.

talldan commented:
Hi @aduth - I've switched over to using a resolver, but there's still a bit left to do. I'll be afk for a couple of weeks, so won't be able to tackle those things for a while. I've left comments about the parts that I think need more work.

noisysocks commented:
I can fill in for Dan while he's AFK 🙂

aduth commented on packages/editor/src/store/resolvers.js:
```diff
@@ -0,0 +1,19 @@
+/**
```
`core-data` is meant to encompass all core entities, which in my mind should encompass autosaves. In other words, I think these resolvers and associated behavior ought to be moved there. Should help simplify things a fair bit as well (avoid reimplementing `SELECT`, `API_FETCH` controls).

talldan replied:
It looks like that would entail moving all the post state as well, since there's some coupling in the selectors:
https://github.com/WordPress/gutenberg/blob/master/packages/editor/src/store/selectors.js#L493

Just want to confirm that's what you're proposing. Either I do that or use a different strategy - which I would imagine means moving the selector logic out into the components that use it.

talldan replied:
I managed to make this work by keeping `isEditedPostAutosaveable` in the editor store and passing autosave into it as an arg.

I still need to sort out deprecations for the old interface.

noisysocks commented on lib/client-assets.php:
```diff
@@ -1264,22 +1291,21 @@ function gutenberg_default_post_format_template( $settings, $post ) {
+	$autosave_response = gutenberg_api_request( sprintf( '/wp/v2/posts/%s/autosaves?context=edit', $post->ID ) );
 
-	if ( ! $autosave ) {
+	if ( ! isset( $autosave_response['body'][0] ) ) {
```
It's because `isset` is a language construct and not a function. Wild, hey!

talldan commented:
@gziolo - it's still WIP. We've been on support rotation this week.

Will start on it again on Monday. The refactoring of autosave state to core/data isn't very straightforward that's what I'm working through at the moment.

adamsilverstein commented on packages/editor/src/store/selectors.js:
```diff
@@ -485,18 +486,29 @@ export function isEditedPostEmpty( state ) {
  */
-export function isEditedPostAutosaveable( state ) {
+export function isEditedPostAutosaveable( state, autosave ) {
+	if ( autosave === undefined ) {
```
Why did you decide to pass the autosave into `isEditedPostAutosaveable` instead of always retrieving it here?

talldan replied:
Selectors should be functionally pure, and I could only achieve that by passing the autosave state in as an argument now that it no longer lives in the same state tree.

adamsilverstein replied:
Makes sense, thanks for clarifying!

talldan commented:
Tests seem to be failing for the custom post type in the templates test. The autosave endpoint returns 404 rest_post_invalid_parent, which seems to trigger a console error and that's what causes the test failure. Working on that now.

talldan commented on packages/core-data/src/resolvers.js:
```diff
@@ -126,14 +126,13 @@ export function* hasUploadPermissions() {
-	const postTypePathPart = type === 'post' ? 'posts' : type;
+	const postTypeEntities = yield getKindEntities( 'postType' );
+	const entity = find( postTypeEntities, { name: type } );
+	const { baseURL } = entity;
```
I have no idea if this is the right way to get the correct `baseURL` for the post type. Some validation or advice on alternatives would be appreciated.

talldan commented on packages/editor/src/store/selectors.js:
```diff
@@ -485,18 +489,29 @@ export function isEditedPostEmpty( state ) {
-export function isEditedPostAutosaveable( state ) {
+export function isEditedPostAutosaveable( state, autosave ) {
+	if ( arguments.length === 1 ) {
+		deprecated( '`wp.data.select( \'core/editor\' ).isEditedPostAutosaveable()`', {
```
I'm not sure if this is the right way to handle a change of function signature.

aduth replied:
> I'm not sure if this is the right way to handle a change of function signature.

I think it's as good as we'll get with deprecation, for the moment.

The main issue here is that the top-level `select` function is not contextualized to a specific data registry, so it's not guaranteed to be operating on the same data as where the state provided to the selector is actually coming from. This is normally provided by `withSelect` using a `RegistryConsumer`:

https://github.com/WordPress/gutenberg/blob/71be7e60faac03624d37d436e91a258b72457131/packages/data/src/components/with-select/index.js#L149-L156

Previously I'd thought we'd want a way to be able to compose selectors across stores, though the idea of having the requirement be to pass data as arguments is also compelling. Selecting across stores is not simple to implement, and most likely wouldn't qualify as being considered "pure".

noisysocks replied:
Andrew is right here that using `select()` will always use the default registry and not any specific registry that is set using `RegistryProvider`. 

@aduth: do you have any suggestions on how to address this?

talldan replied:
I've updated the code to use registrySelectors in https://github.com/WordPress/gutenberg/pull/7945/commits/1a489e2b2dc2d545da3ad6c886e26166f9dd7503, the tests for `isEditedPostAutosaveable` also needed to be changed in https://github.com/WordPress/gutenberg/pull/7945/commits/e349caaab00b0a132cf4820a5b9988aad2ddc469 to take into account that it's a registry selector. I've added comments that hopefully make the situation clear.

Would also note that this PR introduces some use of `select` in an effect, and I haven't converted that to a registry selector. Not entirely sure of the best strategy there, but I note that there's also some existing use of select in effects:
https://github.com/WordPress/gutenberg/blob/master/packages/edit-post/src/store/effects.js#L46-L54

I've tested, the selectors still work, but unsure of how to assert that they work in the context of a different registry 🤔

talldan commented:
This should be good for a re-review. Thanks for the patience!

aduth commented on packages/core-data/src/resolvers.js:
```diff
@@ -118,3 +119,18 @@ export function* hasUploadPermissions() {
+/**
+ * Request autosave data from the REST API.
+ *
+ * @param {Object} post The post that is the parent of the autosave.
```
Question: How much more difficult would it be to implement as accepting the post's ID, not the object? The former is much more accessible, thus making the selector easier to use. Generally I think we should seek to accept simpler arguments. While not relevant here, it also enables easier function memoization, since complex arguments can't be compared using strict equality conditions.

talldan replied:
The post's type is needed to form the correct path to the rest endpoint (wp/v2/:post_type_rest_base/:id/autosaves). I can change the signature to `getAutosave( postType, postId )`. Not necessarily easier for the implementor, but at least simpler types.

aduth replied:
Hm. It's a shame that despite post IDs being unique across all post types, there's no way to retrieve a post from the API knowing only its ID.

I still think the variant you suggest would be preferable, though the ultimate ideal for me would be one which accepted just the ID.

aduth replied:
Posed the question in the `core-restapi` Slack channel:

https://wordpress.slack.com/archives/C02RQC26G/p1546983215113700

I don't think there's currently a way to get the entity with only the ID. There was some contemplation as to whether this could be an addition to the search endpoint.

talldan replied:
The other thing that adds complexity is that the autosave is fetched by its parent's postType and postId rather than its own postId, so that would make it quite a specialised endpoint.

aduth commented on packages/core-data/src/reducer.js:
```diff
@@ -234,6 +234,45 @@ export function hasUploadPermissions( state = true, action ) {
+				'title',
+				'excerpt',
+				'content',
+			].map( ( field ) => {
```
To me, given that the intent of the `core-data` package is to serve as providing the raw underlying data, I don't know that we should be doing any post-processing here. I'd expect an autosave "entity" to be exactly what is provided through the REST API.

talldan replied:
Seems as though there are three main consumers of the data:
- The deprecated selectors.
- The `requestPostUpdate` effect uses `getAutosave`, which would now return a different data structure.
- The `isEditedPostAutosaveable` selector also compares the autosave, passed in from `getAutosave`.

In those places I could use a function to transform the data, `getSavedAutosaveAttributes( autosave )`. I'm not sure where that function should live. Any thoughts?

aduth replied:
Would it be enough to just use the `getPostRawValue` function from the editor store's `reducer.js`?

aduth commented on packages/editor/src/store/selectors.js:
```diff
@@ -321,20 +323,22 @@ export function getEditedPostAttribute( state, attributeName ) {
  *
  * @return {*} Autosave attribute value.
  */
 export function getAutosaveAttribute( state, attributeName ) {
```
I guess this selector is never used anywhere in Gutenberg? A shame then that we included it with 5.0. Per my prior comment, I don't know that we even want a selector like `getAutosaveAttribute` in `core-data`, and rather we could mimic an implementation to serve as compatibility here (supporting only those subset of properties `content`, `excerpt`, `title`).

aduth commented on packages/editor/src/store/selectors.js:
```diff
@@ -485,18 +489,29 @@ export function isEditedPostEmpty( state ) {
  * Returns true if the post can be autosaved, or false otherwise.
  *
- * @param  {Object}  state Global application state.
+ * @param  {Object} state    Global application state.
```
The spacing alignment prior to `{` to align to the `@return` is not necessary or correct, per JavaScript Documentation Guidelines:

https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/#functions

talldan replied:
Fixed

aduth commented on packages/editor/src/store/actions.js:
```diff
@@ -44,15 +46,20 @@ export function resetPost( post ) {
+		plugin: 'Gutenberg',
+	} );
+
+	const postId = select( 'core' ).getCurrentPostId();
```
`'core'` doesn't have `getCurrentPostId`. I assume it should be `'core/editor'`.

talldan replied:
Yep, thanks for catching that mistake.

aduth commented on packages/editor/src/components/post-preview-button/index.js:
```diff
@@ -210,17 +211,23 @@ export default compose( [
 
 		const previewLink = getEditedPostPreviewLink();
 		const postType = getPostType( getEditedPostAttribute( 'type' ) );
+		const post = getCurrentPost();
```
Another reason simple arguments to selectors would be preferable is in considering that a selector like this would could potentially be quite non-performant in aggregating the properties of the object of a "post", whereas getting only its `id` would most certainly be trivial in comparison.

`getCurrentPost` is actually trivial, but something like `getBlock` is a good example of the opposite.

aduth commented on packages/editor/src/components/post-preview-button/index.js:
```diff
@@ -210,17 +211,23 @@ export default compose( [
+		const autosave = getAutosave( post );
+
 		return {
+			autosave,
```
Why is this being passed as a prop now? Especially considering the component already receives `autosave` from `withDispatch` as a _function_, which this is not.

talldan replied:
Yeah, that's a mistake, I've removed it.

aduth commented on packages/editor/CHANGELOG.md:
```diff
@@ -4,6 +4,12 @@
+### Deprecations
+
+- The `getAutosave`, `getAutosaveAttribute`, and `hasAutosave` selectors are deprecated. Equivalent selectors have been added to the `@wordpress/core-data` package.
+- The `resetAutosave` action creator is deprecated. An equivalent action creator `receiveAutosave` has been added to the `@wordpress/core-data` package.
```
I don't know that "action creator" as in the Redux lexicon quite fits for how we expose these things, since they're already prebound to dispatch. As far as public documentation goes, I find it's okay to refer to them as "actions", or perhaps "action dispatchers".

talldan replied:
Have updated this to just 'actions'.

aduth commented on packages/core-data/src/selectors.js:
```diff
@@ -180,3 +193,62 @@ export function isPreviewEmbedFallback( state, url ) {
+ * @return {?Object} The autosave object, or undefined if there is none.
+ */
+export function getAutosave( state, post ) {
+	if ( state.autosave ) {
```
When is `state.autosave` falsey?

talldan replied:
Thanks for catching that, removed the `if`.

talldan commented on packages/core-data/src/selectors.js:
```diff
@@ -180,3 +193,42 @@ export function isPreviewEmbedFallback( state, url ) {
+ * @return {boolean} Whether there is an existing autosave.
+ */
+export function hasAutosave( state, postType, postId ) {
+	return !! getAutosave( state, postType, postId );
```
Accessing another selector in the same file like this won't trigger the resolver, which makes it incorrectly return false when the autosave might simply not have been fetched yet. I'm wondering what the best alternative would be ... maybe just drop this selector in favour of `!! select( 'core' ).getAutosave()`?

talldan replied:
I decided against adding `hasAutosave` to core data, like `getAutosaveAttribute` it wasn't being used after my changes. Deleting it was the simplest solution!

noisysocks commented on packages/editor/src/components/post-preview-button/index.js:
```diff
@@ -210,17 +210,23 @@ export default compose( [
 			previewLink: forcePreviewLink !== undefined ? forcePreviewLink : previewLink,
-			isSaveable: isEditedPostSaveable(),
-			isAutosaveable: forceIsAutosaveable || isEditedPostAutosaveable(),
+			isSaveable: isEditedPostSaveable() && hasFetchedAutosave( postTypeName, postId ),
```
nit: `isEnabled` would be a more accurate name for this prop now.

talldan replied:
Hmm, ok. Would this require a deprecation for the inner component? I'm not sure if the unenhanced `PostPreviewButton` is considered public.

noisysocks replied:
> I'm not sure if the unenhanced PostPreviewButton is considered public.

I'd say no: only the exported enhanced component should be considered our public interface.

noisysocks commented on packages/editor/src/store/effects/posts.js:
```diff
@@ -103,11 +106,14 @@ export const requestPostUpdate = async ( action, store ) => {
 	let request;
 	if ( isAutosave ) {
+		const autosave = dataSelect( 'core' ).getAutosave( post.type, post.id );
+		const mappedAutosave = mapValues( pick( autosave, AUTOSAVE_PROPERTIES ), getPostRawValue );
```
Is the `pick()` here necessary? When would `autosave` have a property that's not listed in `AUTOSAVE_PROPERTIES`?

noisysocks commented on packages/editor/src/store/effects/posts.js:
```diff
@@ -317,3 +326,4 @@ export const refreshPost = async ( action, store ) => {
 	} );
 	dispatch( resetPost( newPost ) );
 };
+
```
nit: Unnecessary new line here.

talldan replied:
Removed.

noisysocks commented on packages/editor/src/store/actions.js:
```diff
@@ -44,15 +46,20 @@ export function resetPost( post ) {
  * post has been received, by initialization or autosave.
  *
- * @param {Object} post Autosave post object.
+ * @param {Object} newAutosave Autosave post object.
```
nit: Should we add a `@deprecated` comment here?

talldan replied:
Definitely. Done.

noisysocks commented on packages/editor/src/store/selectors.js:
```diff
@@ -519,10 +549,9 @@ export function isEditedPostAutosaveable( state ) {
-	// If the title, excerpt or content has changed, the post is autosaveable.
-	const autosave = getAutosave( state );
+	// If the title or excerpt has changed, the post is autosaveable.
 	return [ 'title', 'excerpt' ].some( ( field ) => (
```
Should we use `AUTOSAVE_PROPERTIES` here?

talldan replied:
`AUTOSAVE_PROPERTIES` also has the `content` property, which isn't used on that line, it's compared in the preceding lines in a slightly different fashion. I suppose something like `omit( AUTOSAVE_PROPERTIES, 'content' )` could be used. What do you think?

noisysocks replied:
Eh, leave it as is 🙂

noisysocks commented:
Thanks for keeping on top of this @talldan — I know it's been a slog! ❤️

I followed the steps in the issue locally and wasn't able to reproduce the bug anymore. I commented with a few minor nits here and there but otherwise the code is looking good to me.

Before merging I would like if @aduth could take another quick look at this. In particular I'm not sure how we ought to address the problem outlined in https://github.com/WordPress/gutenberg/pull/7945#discussion_r245782617 which could be an issue for backwards compatibility.

talldan commented on packages/core-data/src/reducer.js:
```diff
@@ -234,6 +234,28 @@ export function hasUploadPermissions( state = true, action ) {
+ *
+ * @return {Object} Updated state.
+ */
+export function autosave( state = {}, action ) {
```
good idea 😄. Renamed.

talldan commented on packages/core-data/src/selectors.js:
```diff
@@ -180,3 +193,29 @@ export function isPreviewEmbedFallback( state, url ) {
 }
+
+/**
+ * Returns the autosave that is a child of the provided post id, if one exists.
```
Another good idea 😄 . Done.

talldan commented:
Thanks for the review @noisysocks. I've addressed most of the feedback and asked some questions on others.

noisysocks commented:
Just waiting on something to come of this:

> Before merging I would like if @aduth could take another quick look at this. In particular I'm not sure how we ought to address the problem outlined in #7945 (comment) which could be an issue for backwards compatibility.

aduth commented:
I don't have a good answer. It's the same issue we're facing in #13088 , and it completely nullifies any benefit we can get from optimizations like in #13177 since the dependencies can't easily be followed.

The closest thing to a solution which has been considered is a way to compose a selector such that it has access to the current registry in its scope. For example:

```js
const isEditedPostAutosaveable = bindRegistry( ( registry ) => ( state ) => {
	const postType = getCurrentPostType( state );
	const postId = getCurrentPostId( state );
	const autosave = registry.select( 'core' ).getAutosave( postType, postId );

	// ...
} );
```

...where something in the actual execution of a selector (data module) would be responsible for providing the argument when called.

aduth commented:
Could you look to see whether what @youknowriad introduced with #13662 could provide some solution for your need?

talldan commented:
Thanks for pointing that out, will take a look.

talldan commented:
Ready for a re-review! Hopefully I've implemented `createRegistrySelector` correctly 😄

adamsilverstein commented:
@talldan This is excellent, thanks for your hard work here and deep dive into autosaves. 

I am curious about the requirement to pass the `postType` to the `getAutosave` selector (apologies if this was already discussed).  I realize the REST API endpoint requires that you know the post type  to construct the correct URL. Still, it feels a bit excessive to expect this in the selector - core's current equivalent is `wp_get_post_autosave` which takes the (parent) post ID (only) and as a developer I would expect this selector to work the same way.

Can `getAutosave` work the same way, taking a parent post ID, figuring out its post type (should be available already in state, right?) to construct the `baseURL`.

nerrad commented on packages/core-data/src/resolvers.js:
```diff
@@ -118,3 +119,18 @@ export function* hasUploadPermissions() {
+ * @param {number} postId   The id of the parent post.
+ */
+export function* getAutosave( postType, postId ) {
+	const { baseURL } = yield select( 'getEntity', 'postType', postType );
```
you could use a `resolveSelect( 'getPostType`, postType )` control like what I have in [this pull](https://github.com/WordPress/gutenberg/pull/13716/files#diff-e66041d992a60b9b29e5b80f84d7b612R23) and that should work.

aduth replied:
> you could use a `resolveSelect( 'getPostType', postType )` control like what I have in [this pull](https://github.com/WordPress/gutenberg/pull/13716/files#diff-e66041d992a60b9b29e5b80f84d7b612R23) and that should work.

@talldan Did you try this? It's not unexpected that `select` might return `undefined` if it hasn't yet been resolved.

talldan replied:
I've made that change now. 

I suppose `yield select( 'getEntity', 'postType', postType );` always worked because of preloading? In that case it definitely makes sense to use resolveSelect 👍

nerrad commented:
Just leaving a note here to connect this ticket with the work I'm doing in https://github.com/WordPress/gutenberg/pull/13716 which moves all post effects into controls/action-generators in the `@wordpress/editor` package.  So if my pull gets merged first, it will impact work in this ticket (and vice versa).

talldan commented:
> Can `getAutosave` work the same way, taking a parent post ID, figuring out its post type (should be available already in state, right?) to construct the `baseURL`?

Hi @adamsilverstein - thanks for the comment. I think we can't assume that the post will be already loaded into state, as this package could be used outside of gutenberg and the caller could be querying just about any post. We might be able to resolve a selector for the post using the advice here:
https://github.com/WordPress/gutenberg/pull/7945#discussion_r259913801

I need to rebase this PR again! I'll do that now.

talldan commented:
Hey @nerrad - Resolving conflicts shouldn't be too bad if your PR is merged first. 

This other PR of mine would be affected, however still a question-mark over whether that one will go ahead.
https://github.com/WordPress/gutenberg/pull/13718

talldan commented:
@adamsilverstein Just following up on this comment:

> Can getAutosave work the same way, taking a parent post ID, figuring out its post type (should be available already in state, right?) to construct the baseURL?

I've realised this isn't possible. The issue is that the REST API doesn't support fetching post objects only by the id, it also requires the postType.

talldan committed:
Update deprecation version numbers

adamsilverstein commented:
> I've realised this isn't possible. The issue is that the REST API doesn't support fetching post objects only by the id, it also requires the postType.

@talldan Right, that makes sense.  I thought maybe you would be able to determine the `postType` from state inside `getAutosave` with a selector (`const postType = getPostType( getEditedPostAttribute( 'type' ) );`). (instead of passing it) I guess that isn't possible here?

talldan commented:
@adamsilverstein my understanding is that core-data shouldn't be calling selectors from `editor`—the `core-data` package is a general WordPress data module so its functionality shouldn't be tied to Gutenberg's.

talldan commented:
This is ready for review again. I believe I've resolved the conflicts with @nerrad's recent PR. Would be good to get some validation that I'm using those controls correctly.

Also a 👍 that I'm using the registrySelectors correctly.

aduth commented on lib/client-assets.php:
```diff
@@ -686,6 +686,7 @@ function gutenberg_editor_scripts_and_styles( $hook ) {
 		sprintf( '/wp/v2/users/me?post_type=%s&context=edit', $post_type ),
 		array( '/wp/v2/media', 'OPTIONS' ),
 		array( '/wp/v2/blocks', 'OPTIONS' ),
+		sprintf( '/wp/v2/%s/%s/autosaves?context=edit', $rest_base, $post->ID ),
```
I dunno how much we care to use typed placeholders, but I see no reason why the second ought not be `%d`. The same could be said for the `sprintf( '/wp/v2/%s/%s?context=edit', $rest_base, $post->ID )` a few lines up, too.

aduth replied:
FYI: All this code will be removed as of #13569 . You should be prepared to submit an upstream patch via Trac. Do we need it to exist in the meantime in Gutenberg? If so, we should start thinking about how we can extend this from the plugin (i.e. the `block_editor_preload_paths` filter).

aduth commented on packages/core-data/src/reducer.js:
```diff
@@ -238,6 +238,28 @@ export function userPermissions( state = {}, action ) {
+			const { postId, autosave } = action;
+
+			return {
+				[ postId ]: autosave,
```
The REST API schema implies that a post may have multiple autosaves, but in this implementation (and with `getAutosave`) we seem to assume one autosave per post. Who's right?

talldan replied:
Looks like it's one per author:
https://github.com/WordPress/wordpress-develop/blob/master/src/wp-includes/rest-api/endpoints/class-wp-rest-autosaves-controller.php#L332

I did some testing and the endpoint can return multiple autosave for a post when edited by multiple users. I'll also have to key by user id, and add that as a parameter to the selector. 

I'm a bit concerned about lack of usability of the selector within the Gutenberg codebase. 

I wonder if it's worth considering keeping the deprecated selectors/actions in the editor module as proxies?

talldan replied:
Struggling to see how I can determine the user id within Gutenberg ...

aduth replied:
> Struggling to see how I can determine the current user's id within Gutenberg ...
> 
> I think I'll need that when using the selector from the editor package.

There is a `/wp/v2/users/me` endpoint which could be made to be part of a selector / resolver pair (or built-in entity).

That seems the purest approach, though arguably a bit wasteful when considering the user ID is more readily available via something like the suffix of the `WP_DATA_USER_` localStorage persistence key (managed by data module), or `user_ID` hidden input whenn meta boxes are present.

talldan replied:
Thanks for the details. It looks like `users/me` might be the best option. On testing it looks like there can be multiple `WP_DATA_USER_` plugins if two users use the same browser.

aduth commented on packages/editor/src/store/selectors.js:
```diff
@@ -495,18 +513,34 @@ export function isEditedPostEmpty( state ) {
  * @return {boolean} Whether the post can be autosaved.
  */
-export function isEditedPostAutosaveable( state ) {
+export const isEditedPostAutosaveable = createRegistrySelector( ( select ) => function( state, autosave ) {
```
With being a registry selector now, is there any reason we should explicitly need to accept `autosave` as an argument now? What's wrong with leaving the implementation as if `arguments.length === 1` were the only option? It's not wrong for the editor to assume a post ID (and thus, select the autosave for that post ID).

aduth commented on packages/editor/src/store/test/selectors.js:
```diff
@@ -1302,7 +1263,16 @@ describe( 'selectors', () => {
 
 	describe( 'isEditedPostAutosaveable', () => {
+		// isEditedPostAutosaveable is only implemented as a registry selector
+		// to handle a deprecation. None of the tests here trigger the deprecation,
```
"None of the tests here trigger the deprecation" reads like a lack of needed coverage. Though, depending on your response to my prior comment about avoiding the deprecation, it may not be necessary.

talldan committed:
Add resolveSelect control to core-data store

talldan committed:
Use resolveSelect to query for postType in getAutosave resolver

talldan committed:
Rework state for autosaves to reflect that multiple autosave can be received (one per user)

talldan committed:
Add state for the current user

talldan committed:
Add a selector/resolver for retrieving an autosave for an individual author

talldan committed:
Handle providing the current user id to the getAutosave selector

talldan committed:
Undeprecate isEditedPostAutosaveable, update tests to handle registrySelector

talldan committed:
Update changelog and deprecation version numbers

talldan committed:
Update docs

talldan committed:
Add JSDocs for currentUser reducer

talldan committed:
Minor renaming of functions/props

talldan commented:
I've updated this PR again. It is definitely feeling quite complicated. Not sure what opportunities there are to simplify things. I think part of the complication is that the underlying REST API isn't ideal, or at least it doesn't mesh well with how selectors/resolvers work.

Note, I'll need to make a change to core for the preloaded paths, but this is still good to review and test.

talldan committed:
Update deprecation version for action

aduth commented on packages/editor/src/components/post-preview-button/test/__snapshots__/index.js.snap:
```diff
@@ -3,7 +3,7 @@
 <ForwardRef(Button)
   className="editor-post-preview"
-  disabled={false}
+  disabled={true}
```
The tests still reference the now-defunct `isSaveable` and should be updated.

talldan replied:
👍I've ended up switching the PostPreviewButton prop to `isDisabled` which I think is tidier. Also updated the tests this time.

talldan replied:
I've now reintroduced `isSaveable` and sorted the tests.

aduth commented on packages/editor/src/components/post-preview-button/index.js:
```diff
@@ -211,15 +211,19 @@ export default compose( [
 			currentPostLink: getCurrentPostAttribute( 'link' ),
 			previewLink: forcePreviewLink !== undefined ? forcePreviewLink : previewLink,
-			isSaveable: isEditedPostSaveable(),
+			isEnabled: isEditedPostSaveable() && hasFetchedAutosaves( postTypeName, postId ),
```
I'm struggling to understand what implication the availability of fetched autosaves has on whether the post preview button should be disabled. At the least, could we include an inline code comment somewhere which explains this?

talldan replied:
The check is to avoid an autosave attempt being made (via the preview button) when the user already has an autosave that the client hasn't fetched. That would cause a situation where the client and the server disagree, and the server can throw an error (basically the same error that this PR is fixing).

On reflection, I don't think the right strategy. I've pushed a change instead to `isEditedPostAutosaveable` so that it returns false if autosaves have not been fetched:
https://github.com/WordPress/gutenberg/blob/542b4ba844de091636625f1216cd243153749328/packages/editor/src/store/selectors.js#L539-L543

Now, if autosaves have not been fetched, clicking the preview button will cause it to fallback to the existing preview link or post link and it won't attempt an autosave during that process.

aduth commented on packages/editor/src/store/test/selectors.js:
```diff
@@ -1306,6 +1267,15 @@ describe( 'selectors', () => {
 
 	describe( 'isEditedPostAutosaveable', () => {
 		it( 'should return false if the post is not saveable', () => {
+			const isEditedPostAutosaveable = isEditedPostAutosaveableRegistrySelector( () => ( {
```
Was there some prior art for this convention? It seems reasonable, albeit a bit awkward; the nature of these registry selectors I'd assume.

talldan replied:
I couldn't see any. It doesn't seem so bad, but yeah, just a bit awkward.

aduth commented on packages/editor/src/store/actions.js:
```diff
@@ -330,15 +347,16 @@ export function* savePost( options = {} ) {
-			'getAutosave',
-		);
+		const currentUser = yield resolveSelect( 'core', 'getCurrentUser' );
+		const currentUserId = currentUser ? currentUser.id : null;
```
How well does `getAutosave` and the following logic handle the case that this value would be `null`?

talldan replied:
I've changed it to `undefined` instead of `null` as `getAutosave` returns early with undefined if no userId is provided.

If `autosave` is also `undefined`, lodash `pick` coerces it back into an object, at which point all of the resulting operations continue to work as expected.

aduth replied:
> If `autosave` is also `undefined`, lodash `pick` coerces it back into an object, at which point all of the resulting operations continue to work as expected.

There's a fair few number of assumptions here to not have a corresponding unit test or at least some inline code comment outlining the expectations. For example, if someone were to naively replace the `pick` implementation with something different in the future.

aduth commented on packages/editor/CHANGELOG.md:
```diff
@@ -15,6 +15,8 @@
 ### Deprecations
 
 - `EditorGlobalKeyboardShortcuts` has been deprecated in favor of `VisualEditorGlobalKeyboardShortcuts`.
+- The `getAutosave`, `getAutosaveAttribute`, and `hasAutosave` selectors are deprecated. Please use the `getAutosave` selector in the `@wordpress/core-data` package.
```
These notes are under a release which has already been published. They should be moved up to the notes for the pending unreleased version.

aduth reviewed (commented):
This seems very much on the right track. I agree with your point about the new interface of `getAutosave( postType, postId, userId )` being rather complex. However, I think a strong case can be made that it fits well in the scope of a selector to produce useful derived data; the usefulness of getting the singular autosave entry for a given user, the derivation source being the results of the autosaves endpoint. The main criticism could be that if it doesn't map directly to an endpoint or set of query arguments for an endpoint, which is probably what we should aim for with the `core-data` package. But, importantly, it feels like one that _could_ exist (e.g. `?author=1` query argument to align with `/posts`).

The alternative I'd see is moving some of this opinionated behavior back into the editor store, where the editor can make more assumptions about a current user to use in deriving the autosaves data. I don't really think it's necessary, however, and appreciate what we've been able to split out of the editor store as a result of these changes.

There's a few review notes I've left which could use some attention. I still need to take the branch for a spin in the browser as well.

talldan committed:
Change PostPreviewButton prop back to isSaveable.

talldan committed:
Do not consider a post autosaveable if existing autosaves have not yet been fetched from the REST API

talldan committed:
Use undefined as the fallback when no currentUser is fetched

talldan committed:
Move changelog entries to unreleased section, and bump minor instead of patch version

aduth commented on packages/editor/src/store/actions.js:
```diff
@@ -362,9 +380,12 @@ export function* savePost( options = {} ) {
 
-		yield dispatch( STORE_KEY, resetAction, newPost );
+		if ( isAutosave ) {
+			yield dispatch( 'core', 'receiveAutosaves', post.id, newPost );
```
`receiveAutosaves` expects an array of autosaves, but we're passing the singular `newPost`. Will that work?

**Edit:**  Yeah, I think this is not working correctly, as it changes the structure of `autosaves[ postId ]` from an array to an object:

![image](https://user-images.githubusercontent.com/1779930/56161897-b4882600-5f98-11e9-80ec-ab14bd209b04.png)

talldan replied:
Great spot. I've made `receiveAutosaves` so that it can accept either an array or a singular object. Added a test case for the conversion from single autosave to an array.

aduth reviewed (commented):
This is looking good. I suspect my note about `receiveAutosaves` needs to be corrected. Should be as simple as changing `newPost` to `[ newPost ]` but we might consider improving the tests in a way which would have covered this issue.

talldan committed:
Allow single or multiple autosaves to be passed to receiveAutosaves

talldan committed:
Update deprecation versions

aduth committed:
Use REDUCER_KEY as arg to select

Co-Authored-By: talldan <daniel.p.richards@gmail.com>

talldan commented:
@aduth - have fixed that issue with `receiveAutosaves`. I've also updated the deprecation versions, so hopefully everything is good 👍

talldan committed:
Update docs

aduth commented:
>This request is preloaded (will require a core patch).

Is this still the case? I don't see any PHP file edits included.

Is it possible that due to an earlier rebase where we'd been relying on redundant code in `client-assets.php` to specify the preload paths, it was removed in a rebase? 

If we need it, we should filter [`block_editor_preload_paths`](https://developer.wordpress.org/reference/hooks/block_editor_preload_paths/).

aduth reviewed (approved):
Thanks for the patience in delayed reviews. Nice work on this one 👍 

Consider my prior comment as far as whether we should be preloading this path. I don't consider it a blocker since the preloading is more an optimization, but I'll leave it to you whether we should include it.

talldan commented:
Thanks so much for you guidance, @aduth! This one has been through a few iterations, but feels good to finally get it merged.

I'll work on some follow-ups for the preloading.

noisysocks commented:
10 months and 5 days—that must be a record! 😀👏

talldan commented:
@aduth here's a ticket with a patch for adding the preloading:
https://core.trac.wordpress.org/ticket/46974

aduth commented:
> @aduth here's a ticket with a patch for adding the preloading:
> https://core.trac.wordpress.org/ticket/46974

Depending if you think it's important, we could introduce a filter in the Gutenberg plugin ahead of this being patched in core.

talldan commented:
> Depending if you think it's important, we could also introduce a filter in the Gutenberg plugin ahead of this being patched in core.

I'm not completely convinced it's needed, but I made a PR - https://github.com/WordPress/gutenberg/pull/15067
